cmd/common-main.go

```diff
@@ -51,22 +51,21 @@ func initConfig() {
 }
 
 func handleCommonCmdArgs(ctx *cli.Context) {
-	// Set configuration directory.
-	{
-		// Get configuration directory from command line argument.
-		configDir := ctx.String("config-dir")
-		if !ctx.IsSet("config-dir") && ctx.GlobalIsSet("config-dir") {
```

Comment:
Use better logic
```go
configDir := ctx.String("config-dir")
if configDir == "" {
    configDir = ctx.GlobalString("config-dir")
    if configDir == "" {
        configDir = ctx.Parent().GlobalString("config-dir")
    }
}
```

This way, it is clear to handle precedence.  If `minio --config-dir=d1 gateway --config-dir=d2 s3 --config-dir=d3`, `d3` is used as first priority.

Reply:
The other option is to error out if the user provides different options for --config-dir, the only way to do it for both gateway and server is:
```
        configDir := ctx.String("config-dir")
        if ctx.Parent().String("config-dir") != "" {
                if configDir != "" {
                        fatalIf(fmt.Errorf("please provide only one config directory"), "")
                }
                configDir = ctx.Parent().String("config-dir")
        }
        if ctx.Parent().Parent() != nil {
                // ctx.Parent().Parent() is valid for gateway mode only.
                if ctx.Parent().Parent().String("config-dir") != "" {
                        if configDir != "" {
                                fatalIf(fmt.Errorf("please provide only one config directory"), "")
                        }
                        configDir = ctx.Parent().Parent().String("config-dir")
                }
        }
        if configDir == "" {
                // Get $HOME/.minio
                configDir = getDefaultConfigDir()
        }
        if configDir == "" {
                fatalIf(fmt.Errorf("missing option"), "Configuration directory cannot be empty")
        }
```

I think we can just do what bala has suggested here to keep it simple (i.e not error out). I tested that it works for both gateway and server.